// @decorators/di/lib/src/container.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Container = void 0;
const errors_1 = require("./errors");
const constants_1 = require("./constants");
const helpers_1 = require("./helpers");
const injection_token_1 = require("./injection-token");
class Container {
    constructor() {
        this.providers = new Map();
    }
    setParent(container) {
        this.parentContainer = container;
    }
    /**
     * Registers new or replaces providers
     */
    provide(providers) {
        this.registerSingleProviders(providers.filter(provider => !provider.multi));
        this.registerMultiProviders(providers.filter(provider => provider.multi));
    }
    /**
     * Gets or instantiates an injectable
     */
    async get(injectable) {
        return this.resolve(injectable);
    }
    /**
     * Verifies that injectable is provided within the scope
     */
    has(injectable) {
        var _a;
        return this.providers.has(injectable) || ((_a = this.parentContainer) === null || _a === void 0 ? void 0 : _a.has(injectable));
    }
    /**
     * Resolves the provider, following deps are resolved recursively
     */
    async resolve(injectable, optional = false, sequenceDeps = []) {
        var _a, _b, _c, _d, _e;
        let provider = this.providers.get(injectable);
        if (!provider && ((_a = this.parentContainer) === null || _a === void 0 ? void 0 : _a.has(injectable))) {
            return this.parentContainer.get(injectable);
        }
        if (!provider && !optional) {
            throw new errors_1.MissingProviderError(injectable);
        }
        provider = provider !== null && provider !== void 0 ? provider : {};
        if (provider.value) {
            return provider.value;
        }
        const resolvedDeps = [];
        for (const [index, dep] of provider.deps.entries()) {
            if (!dep.id) {
                throw new errors_1.InvalidDependencyError((_b = provider.type) !== null && _b !== void 0 ? _b : injectable, provider.deps, index);
            }
            if (sequenceDeps.find(seqDep => seqDep === dep)) {
                throw new errors_1.RecursiveDependencyError((_c = provider.type) !== null && _c !== void 0 ? _c : injectable, provider.deps, index);
            }
            const depProvider = this.providers.get(dep.id);
            if (depProvider) {
                resolvedDeps.push(await this.resolve(dep.id, dep.optional, [...sequenceDeps, ...provider.deps]));
                continue;
            }
            if ((_d = this.parentContainer) === null || _d === void 0 ? void 0 : _d.has(dep.id)) {
                resolvedDeps.push(await this.parentContainer.get(dep.id));
                continue;
            }
            if (!dep.optional) {
                throw new errors_1.MissingDependencyError((_e = provider.type) !== null && _e !== void 0 ? _e : injectable, provider.deps, index);
            }
            resolvedDeps.push(undefined);
        }
        let value;
        if (provider.factory) {
            value = await provider.factory(...resolvedDeps);
        }
        if (provider.type) {
            value = new provider.type(...resolvedDeps);
        }
        this.providers.set(injectable, Object.assign(Object.assign({}, provider), { value }));
        return value;
    }
    registerMultiProviders(providers) {
        providers.forEach(provider => {
            const injectableId = new injection_token_1.InjectionToken((0, helpers_1.injectableToString)(provider.provide));
            // creates dependency provider for the main provider
            this.registerSingleProviders([Object.assign(Object.assign({}, provider), { provide: injectableId })]);
            const containerProvider = this.providers.get(provider.provide);
            if (containerProvider) {
                return containerProvider.deps.push({ id: injectableId });
            }
            // creates main provider that will hold all other providers as deps
            this.registerSingleProviders([{
                    provide: provider.provide,
                    useFactory: (...deps) => deps,
                    deps: [injectableId],
                }]);
        });
    }
    registerSingleProviders(providers) {
        providers.forEach(provider => {
            if (provider.useClass) {
                return this.registerClassProvider(provider);
            }
            if (provider.useFactory) {
                return this.registerFactoryProvider(provider);
            }
            if (provider.useValue) {
                return this.registerValueProvider(provider);
            }
            if (provider.useExisting) {
                return this.registerExistingProvider(provider);
            }
        });
    }
    registerClassProvider(provider) {
        var _a, _b;
        const ids = (_a = Reflect.getMetadata(constants_1.DEP_IDS_METADATA, provider.useClass)) !== null && _a !== void 0 ? _a : [];
        const optionals = (_b = Reflect.getMetadata(constants_1.OPTIONAL_DEPS_METADATA, provider.useClass)) !== null && _b !== void 0 ? _b : [];
        this.providers.set(provider.provide, {
            type: provider.useClass,
            deps: ids.map((id, index) => ({ id, optional: optionals[index] })),
        });
    }
    registerFactoryProvider(provider) {
        var _a, _b;
        this.providers.set(provider.provide, {
            deps: (_b = (_a = provider.deps) === null || _a === void 0 ? void 0 : _a.map(id => ({ id }))) !== null && _b !== void 0 ? _b : [],
            factory: provider.useFactory,
        });
    }
    registerValueProvider(provider) {
        this.providers.set(provider.provide, {
            value: provider.useValue,
        });
    }
    registerExistingProvider(provider) {
        this.registerFactoryProvider({
            provide: provider.provide,
            deps: [provider.useExisting],
            useFactory: (existing) => existing,
        });
    }
}
exports.Container = Container;
